src/raw_rwlock.rs

```diff
@@ -652,114 +652,118 @@ impl RawRwLock {
+                        Ok(_) => return true,
+                        Err(x) => *state = x,
                     }
+                }
 
-                    if self
-                        .state
-                        .compare_exchange_weak(
-                            *state,
-                            state
-                                .checked_add(ONE_READER)
-                                .expect("RwLock reader count overflow"),
-                            Ordering::Acquire,
-                            Ordering::Relaxed,
-                        )
-                        .is_ok()
-                    {
-                        return true;
+                // This is the same condition as try_lock_shared_fast
+                if *state & WRITER_BIT != 0 {
+                    if !recursive || *state & READERS_MASK == 0 {
+                        return false;
                     }
+                }
 
-                    // If there is high contention on the reader count then we want
-                    // to leave some time between attempts to acquire the lock to
-                    // let other threads make progress.
-                    spinwait_shared.spin_no_yield();
-                    *state = self.state.load(Ordering::Relaxed);
+                if self
+                    .state
+                    .compare_exchange_weak(
+                        *state,
+                        state
+                            .checked_add(ONE_READER)
+                            .expect("RwLock reader count overflow"),
+                        Ordering::Acquire,
+                        Ordering::Relaxed,
+                    )
+                    .is_ok()
+                {
+                    return true;
                 }
-            },
-            |state| state & WRITER_BIT != 0,
-        )
+
+                // If there is high contention on the reader count then we want
+                // to leave some time between attempts to acquire the lock to
+                // let other threads make progress.
+                spinwait_shared.spin_no_yield();
+                *state = self.state.load(Ordering::Relaxed);
+            }
+        };
+        let validate = |state| state & WRITER_BIT != 0;
+        // SAFETY:
+        // * `validate` does not panic or call into any function in `parking_lot`.
+        unsafe { self.lock_common(timeout, TOKEN_SHARED, try_lock, validate) }
     }
 
     #[cold]
     fn unlock_shared_slow(&self) {
         // At this point WRITER_PARKED_BIT is set and READER_MASK is empty. We
         // just need to wake up a potentially sleeping pending writer.
+        // Using the 2nd key at addr + 1
+        let addr = self as *const _ as usize + 1;
+        let callback = |result: UnparkResult| {
+            // Clear the WRITER_PARKED_BIT here since there can only be one
+            // parked writer thread.
+            debug_assert!(!result.have_more_threads);
```

Comment:
Seems like for debug builds, we don't uphold the safety requirement to never panic in `callback`. This seems bad. We don't want UB in debug builds.

Reply:
If we really want to make this sanity check here, and we can't panic, maybe we need to resort to aborting the process.

Reply:
Well this assert never triggers in practice, so it should be fine. These kinds of debug asserts are for debugging issues in parking_lot itself rather than for anything depending on user input.

Reply:
Won't the assert they be compiled into any project depending on `parking_lot` that does not build with `--release`? So it's present for anyone debugging anything, not just `parking_lot` itself? 🤔

Reply:
But it doesn't matter since that code is literally unreachable. I only use debug asserts as a way to document invariants, not because I actually expect them to panic.

Reply:
If I depend on `parking_lot` in my crate and build a debug build. I'm still counting on it being being 100% memory safe.

Reply:
I wrote that last comment before my browser had loaded your previous comment..

So. By "literally unreachable" you mean the assert will only panic if `parking_lot` has a bug? Yeah, sure. I don't know. Still does not uphold the safety requirements, if the safety requirements are correctly stated that is.

Reply:
How about creating a feature (that is disabled by default) and an internal version of `debug_assert!` that does nothing when the feature is disabled? Then the documentational purposes of these asserts remain the same, but we stop breaking guarantees.

Reply:
Just remove the assert then if you feel that strongly about it.

Reply:
Done.
But then I started thinking. Is it really unsafe to panic in this `callback`? if `unpark_one` is aborted at the point where it calls `callback` it will leave the bucket for this key locked forever. That will of course cause a deadlock. But deadlocks are not unsafe. Can something else happen if it panics that leads to UB?